tests/ut/test_index_check.cc

```diff
@@ -267,17 +221,37 @@ TEST_CASE("Test index has raw data", "[IndexHasRawData]") {
+            IndexEnum::INDEX_FAISS_IDMAP, min_ver, knowhere::Json::parse(R"({"metric_type": "COSINE"})")));
+        CHECK(knowhere::IndexStaticFaced<fp32>::HasRawData(IndexEnum::INDEX_FAISS_SCANN, ver, json));
 
         json[indexparam::WITH_RAW_DATA] = false;
-        CHECK_FALSE(KnowhereCheck::IndexHasRawData<fp32>(IndexEnum::INDEX_FAISS_SCANN, metric::L2, ver, json));
+        CHECK_FALSE(knowhere::IndexStaticFaced<fp32>::HasRawData(IndexEnum::INDEX_FAISS_SCANN, ver, json));
         json[indexparam::WITH_RAW_DATA] = "true";
-        CHECK(KnowhereCheck::IndexHasRawData<fp32>(IndexEnum::INDEX_FAISS_SCANN, metric::L2, ver, json));
+        CHECK(knowhere::IndexStaticFaced<fp32>::HasRawData(IndexEnum::INDEX_FAISS_SCANN, ver, json));
         json[indexparam::WITH_RAW_DATA] = "false";
-        CHECK_FALSE(KnowhereCheck::IndexHasRawData<fp32>(IndexEnum::INDEX_FAISS_SCANN, metric::L2, ver, json));
+        CHECK_FALSE(knowhere::IndexStaticFaced<fp32>::HasRawData(IndexEnum::INDEX_FAISS_SCANN, ver, json));
         json[indexparam::WITH_RAW_DATA] = 1;
-        CHECK_FALSE(KnowhereCheck::IndexHasRawData<fp32>(IndexEnum::INDEX_FAISS_SCANN, metric::L2, ver, json));
+        CHECK_FALSE(knowhere::IndexStaticFaced<fp32>::HasRawData(IndexEnum::INDEX_FAISS_SCANN, ver, json));
+
+        knowhere::Json faiss_hnsw_cfg = {{"refine", true}, {"refine_type", "bf16"}};
+        CHECK_FALSE(knowhere::IndexStaticFaced<fp32>::HasRawData(IndexEnum::INDEX_FAISS_HNSW_SQ, ver, faiss_hnsw_cfg));
+        CHECK(knowhere::IndexStaticFaced<bf16>::HasRawData(IndexEnum::INDEX_FAISS_HNSW_SQ, ver, faiss_hnsw_cfg));
+        CHECK_FALSE(knowhere::IndexStaticFaced<fp16>::HasRawData(IndexEnum::INDEX_FAISS_HNSW_SQ, ver, faiss_hnsw_cfg));
+
+        faiss_hnsw_cfg["refine_type"] = "fp16";
+        CHECK_FALSE(knowhere::IndexStaticFaced<fp32>::HasRawData(IndexEnum::INDEX_FAISS_HNSW_SQ, ver, faiss_hnsw_cfg));
+        CHECK_FALSE(knowhere::IndexStaticFaced<bf16>::HasRawData(IndexEnum::INDEX_FAISS_HNSW_SQ, ver, faiss_hnsw_cfg));
+        CHECK(knowhere::IndexStaticFaced<fp16>::HasRawData(IndexEnum::INDEX_FAISS_HNSW_SQ, ver, faiss_hnsw_cfg));
+
+        faiss_hnsw_cfg["refine_type"] = "sq8";
+        CHECK_FALSE(knowhere::IndexStaticFaced<fp32>::HasRawData(IndexEnum::INDEX_FAISS_HNSW_SQ, ver, faiss_hnsw_cfg));
+        CHECK_FALSE(knowhere::IndexStaticFaced<bf16>::HasRawData(IndexEnum::INDEX_FAISS_HNSW_SQ, ver, faiss_hnsw_cfg));
+        CHECK_FALSE(knowhere::IndexStaticFaced<fp16>::HasRawData(IndexEnum::INDEX_FAISS_HNSW_SQ, ver, faiss_hnsw_cfg));
+
+        faiss_hnsw_cfg["refine_type"] = "sq6";
+        CHECK_FALSE(knowhere::IndexStaticFaced<fp32>::HasRawData(IndexEnum::INDEX_FAISS_HNSW_SQ, ver, faiss_hnsw_cfg));
+        CHECK_FALSE(knowhere::IndexStaticFaced<bf16>::HasRawData(IndexEnum::INDEX_FAISS_HNSW_SQ, ver, faiss_hnsw_cfg));
+        CHECK_FALSE(knowhere::IndexStaticFaced<fp16>::HasRawData(IndexEnum::INDEX_FAISS_HNSW_SQ, ver, faiss_hnsw_cfg));
```

Comment:
technically, it is a bit more complicated.
According to my unit tests, 
* `INDEX_FAISS_SQ` `SQ6` accepts `SQ8` refines for any datatype. BF16/FP16/FP32 are correct.
* `INDEX_FAISS_HNSW_PQ` accepts `SQ6` and `SQ8` for any input data type
* `INDEX_FAISS_HNSW_PRQ` accepts `SQ6` and `SQ8` for any input data type
Would it be possible to implement such a check? Or is it way too specialized at this stage?

Reply:
It is a little difficult to comprehensive converage the faiss_hnsw here, with many related permutations and combinations. The ut here is flatly spread out. In fact, the comparision between static and dynamic interfaces has been added. https://github.com/zilliztech/knowhere/pull/836/files/e8b75cacbbb42ea549dd72489117e156be97fdec#diff-e80c39834b0d7812c961f173166720c2111791f5920d79ed686bd7944679fd51R199